lib/TPP/Transforms/LoopInsertion.cpp

```diff
@@ -0,0 +1,307 @@
+  tileShapeM.push_back(mTileShape);
+  tileShapeM.push_back(mShape[0] / mTileShape);
+
+  SmallVector<unsigned> tileShapeN;
+  tileShapeN.push_back(nTileShape);
+  tileShapeN.push_back(nShape[0] / nTileShape);
+
+  // Validate the input tile sizes against the operand shapes
+  long multipleM = 1;
+  for (size_t i = 0; i < tileShapeM.size(); i++)
+    multipleM = multipleM * tileShapeM[i];
+
+  if (mShape[0] != multipleM)
+    return rewriter.notifyMatchFailure(
+        op, "require m tile shape to match tensor shape");
+
+  long multipleN = 1;
+  for (size_t i = 0; i < tileShapeN.size(); i++)
+    multipleN = multipleN * tileShapeN[i];
+
+  if (nShape[0] != multipleN)
+    return rewriter.notifyMatchFailure(
+        op, "require n tile shape to match tensor shape");
+
+  if ((multipleM * multipleN) != (kShape[0] * kShape[1]))
+    return rewriter.notifyMatchFailure(
+        op, "require k tile shape to match tensor shape");
+
+  int boundSize = tileShapeM.size() + tileShapeN.size();
+
+  // Set the new bounds of for loop
+  SmallVector<int64_t> lbs(boundSize, 0), steps(boundSize, 1);
+
+  SmallVector<int64_t> ubs(tileShapeM.begin(), tileShapeM.end());
+  ubs.append(tileShapeN.begin(), tileShapeN.end());
+
+  op.setStaticLowerBound(lbs);
+  op.setStaticUpperBound(ubs);
+  op.setStaticStep(steps);
+
+  // Add new induction var args to the for loop
+  int numArgs = op.getBody()->getArguments().size();
+
+  for (int i = 0; i < boundSize - numArgs; i++)
+    op.getBody()->addArgument(b.getIndexType(), op.getLoc());
+
+  SmallVector<int64_t> tileOffsets{
+      0, static_cast<int64_t>(tileShapeM.size()),
+      static_cast<int64_t>(tileShapeN.size() + tileShapeM.size())};
+  rewriter.setInsertionPointToStart(op.getBody());
+
+  SmallVector<ArrayRef<int64_t>> originalShapes{mShape, nShape, kShape};
+  SmallVector<SmallVector<SmallVector<unsigned>>> tilingVectors{
+      {tileShapeM}, {tileShapeN}, {tileShapeM, tileShapeN}};
+
+  // Replace old args with newly computed args
+  for (auto oper = op.getBody()->getOperations().begin();
+       oper != op.getBody()->getOperations().end(); oper++) {
+    int operandIndex = 0;
+    for (auto arg : oper->getOperands()) {
+      int oldArgIndex = -1;
+      for (int i = 0; i < numArgs; i++) {
+        if (arg == op.getInductionVar(i)) {
+          oldArgIndex = i;
+          break;
+        }
+      }
+      if (oldArgIndex != -1) {
+        Value add = op.getBody()->getArgument(tileOffsets[oldArgIndex]);
+        Value mul;
+        for (int j = tileOffsets[oldArgIndex] + 1;
+             j < tileOffsets[oldArgIndex + 1]; j++) {
+          unsigned cumulativeUpperBound = 1;
+          for (int k = j; k < tileOffsets[oldArgIndex + 1]; k++)
+            cumulativeUpperBound *= op.getStaticUpperBound()[j];
+          Value upperBound = rewriter.create<arith::ConstantIndexOp>(
+              op.getLoc(), cumulativeUpperBound);
+          mul = rewriter.create<arith::MulIOp>(op.getLoc(), b.getIndexType(),
```

Comment:
Let's not introduce indexing arithmetic where we can just use subviews, expand shape and tiling.

Reply:
I have added expand shape wherever possible, however there are affine maps which use induction variables, which cannot be updated to use expand shapes in case of torch mlp model for example. We need mul-add chains in such cases as fallback. I have simplified this code as much as possible so that it stays bug free. It now uses getStridesAndOffset call to compute strides and computes and mul-add chain of induction variables. Please take a look.